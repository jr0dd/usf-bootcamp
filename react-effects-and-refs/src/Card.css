.Card {
  margin: auto;
  position: absolute;
}