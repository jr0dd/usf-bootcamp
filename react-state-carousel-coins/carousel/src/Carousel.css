.Carousel-main {
  display: flex;
  align-items: center;
  justify-content: center;
}

.Carousel-main > i {
  cursor: pointer;
  margin: 10px;
  color: #aaa;
}

.Carousel-main > i:hover {
  color: black;
}

.bi {
  font-size: 35px
}
