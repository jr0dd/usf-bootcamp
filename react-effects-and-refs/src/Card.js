import React from 'react'
import './Card.css'

const Card = ({ name, image }) => {
  return <img
    className='Card'
    alt={name}
    src={image}
  />
}

export default Card