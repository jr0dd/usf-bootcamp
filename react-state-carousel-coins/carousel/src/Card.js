import React from "react";
import "./Card.css";

/** Card: displays image.
 * 
 * Props:
 * - caption: string describing the image
 * - src: string for the image link
 * - currNum: integer for what image we're on
 * - totalNum: integer for how many images are in the collection
 * 
 * State:
 * - none
 * 
 * App --> Carousel --> Card
 */

 function Card({caption, src, currNum, totalNum}) {

  return (
    <div className="Card">
      <h4 className="Card-title">{caption}</h4>
      <img className="Card-image" src={src} alt={caption} />
      <small className="Card-small">
        Image {currNum} of {totalNum}.
      </small>
    </div>
  );
}

export default Card;
