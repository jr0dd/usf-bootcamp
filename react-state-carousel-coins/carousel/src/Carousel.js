import { useState } from 'react'
import './Carousel.css'
import Card from './Card'

/** Carousel: displays images and arrows to navigate through them
 *
 * Props:
 * - photos: array of {src, caption} objects
 * - title: string describing the collection of images
 *
 * State:
 * - currCardIdx: integer for current card index
 *
 * App --> Carousel --> Card
 */
function Carousel ({ photos, title }) {
  const [currCardIdx, setCurrCardIdx] = useState(0)

  const currCard = photos[currCardIdx]
  const total = photos.length

  // Increments currCardIdx state by 1
  function goForward () {
    setCurrCardIdx(currCardIdx + 1)
  }

  // Increments currCardIdx state by 1
  function goBackward () {
    setCurrCardIdx(currCardIdx - 1)
  }

  return (
    <div className='Carousel'>
      <h1>{title}</h1>
      <div className='Carousel-main'>
        {currCardIdx !== 0
        ?  (<i
            className='bi bi-arrow-left-circle'
            onClick={goBackward}
          />)
        : (<i></i>)
        }

      
        <Card
          caption={currCard.caption}
          src={currCard.src}
          currNum={currCardIdx + 1}
          totalNum={total}
        />
        {currCardIdx !== total - 1 
        ? (<i
          className='bi bi-arrow-right-circle'
          onClick={goForward}
        />)
        : (<i></i>)}
      </div>
    </div>
  )
}

export default Carousel
