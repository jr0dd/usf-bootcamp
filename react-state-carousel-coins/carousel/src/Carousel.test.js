import { render, fireEvent } from '@testing-library/react'
import Carousel from './Carousel'
import Card from './Card'
import TEST_IMAGES from './_testCommon.js'

// smoke
it('renders the carousel', () => {
  render(
    <Carousel
      photos={TEST_IMAGES}
      title='images for testing'
    />
  )
})

// snapshot
it('matches snapshot', () => {
  const { asFragment } =   render(
    <Carousel
      photos={TEST_IMAGES}
      title='images for testing'
    />
  )
  expect(asFragment()).toMatchSnapshot()
})

it('works when you click on the right arrow', () => {
  const { container } = render(
    <Carousel
      photos={TEST_IMAGES}
      title='images for testing'
    />
  )
  // expect the first image to show, but not the second
  expect(
    container.querySelector('img[alt="testing image 1"]')
  ).toBeInTheDocument()
  expect(
    container.querySelector('img[alt="testing image 2"]')
  ).not.toBeInTheDocument()

  // move forward in the carousel
  const rightArrow = container.querySelector('.bi-arrow-right-circle')
  fireEvent.click(rightArrow)

  // expect the second image to show, but not the first
  expect(
    container.querySelector('img[alt="testing image 1"]')
  ).not.toBeInTheDocument()
  expect(
    container.querySelector('img[alt="testing image 2"]')
  ).toBeInTheDocument()
})

// test left arrow
it('works when you click on the left arrow', () => {
  const { container } = render(
    <Carousel
      photos={TEST_IMAGES}
      title='images for testing'
    />
  )

  // expect the first image to show, but not the second
  expect(
    container.querySelector('img[alt="testing image 1"]')
  ).toBeInTheDocument()
  expect(
    container.querySelector('img[alt="testing image 2"]')
  ).not.toBeInTheDocument()


  // move forward in the carousel so arrow displays
  const rightArrow = container.querySelector('.bi-arrow-right-circle')
  fireEvent.click(rightArrow)

  // move backward in the carousel
  const leftArrow = container.querySelector('.bi-arrow-left-circle')
  fireEvent.click(leftArrow)

  // expect the first image to show, but not the second
  expect(
    container.querySelector('img[alt="testing image 2"]')
  ).not.toBeInTheDocument()
  expect(
    container.querySelector('img[alt="testing image 1"]')
  ).toBeInTheDocument()
})

it('left arrow missing on first image', () => {
  const { container } = render(
    <Carousel
      photos={TEST_IMAGES}
      title='images for testing'
    />
  )

  expect(
    container.querySelector('.bi-arrow-left-circle')
  ).not.toBeInTheDocument()
})

it('right arrow missing on last image', () => {
  const { container } = render(
    <Carousel
      photos={TEST_IMAGES}
      title='images for testing'
    />
  )

  const rightArrow = container.querySelector('.bi-arrow-right-circle')
  for (let i = 0; i < TEST_IMAGES.length; i++) {
    fireEvent.click(rightArrow)
  }

  expect(
    container.querySelector('.bi-arrow-right-circle')
  ).not.toBeInTheDocument()
})