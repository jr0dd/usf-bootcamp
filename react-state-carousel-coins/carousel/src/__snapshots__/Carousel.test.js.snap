// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`matches snapshot 1`] = `
<DocumentFragment>
  <div
    class="Carousel"
  >
    <h1>
      images for testing
    </h1>
    <div
      class="Carousel-main"
    >
      <i />
      <div
        class="Card"
      >
        <h4
          class="Card-title"
        >
          testing image 1
        </h4>
        <img
          alt="testing image 1"
          class="Card-image"
          src="test1.com"
        />
        <small
          class="Card-small"
        >
          Image 1 of 3.
        </small>
      </div>
      <i
        class="bi bi-arrow-right-circle"
      />
    </div>
  </div>
</DocumentFragment>
`;
